### ISSUE TITLE
If possible, deploy EKS pods to a private subnet, not public

### ISSUE THREAD

**Issue by kostaleonard** [kostaleonard/populare-iac#21]
We want to deploy our pods to a private subnet if possible. According to [AWS](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Scenario2.html), it seems that the instances in a public subnet can send outbound traffic directly to the internet, whereas the instances in a private subnet can't. Currently, we deploy our pods to a public subnet because we want Internet users to be able to reach the reverse proxy pods, web server pods, and database proxy pods. However, having the entire EKS cluster in public subnets exposes many other services with which we do not intend users to interact.

[This AWS EKS networking guide](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html) states that only the load balancer need be in a public subnet; all other nodes can be in private subnets.

**Comment by kostaleonard** [comment]
This is a good idea, but it does require the implementation of significant infrastructure to support. Because instances in a private subnet cannot reach the Internet, all configuration must happen before any resources are deployed. In particular, all apps must be containerized (which they already are in our setup), and those containers must be stored in an AWS container registry. In addition, we would also need to mirror certain packages or find other ways to make them installable, e.g., when we start the Wireguard EC2 node, it needs to `apt install docker.io` and its dependencies.

I am moving this to second release as possible future work.

**Comment by kostaleonard** [comment]
Based on some research, I believe that the following microservices should be in the public subnets:

* Web app
* Database proxy
* Reverse proxy
* VPN

The reason is that the Flutter web app does not (and, as far as I can tell, cannot) perform any server-side computation. Flask apps perform server-side computation, so if I had used Flask for the web server, I could have made the request on the database without needing a proxy. Since Flutter cannot do that, I need a database proxy that a user can reach to execute code that will interface with the database from within the VPC. The reverse proxy is also necessary so that both the web app and the database proxy can be accessed from a single load balancer URL--the database proxy Kubernetes Service could have been a LoadBalancer Service instead, but then we would be incurring extra costs and have another entry point into our VPC. Finally, the VPN needs to be in a public subnet so that admins can reach it from the Internet.

All other microservices, including Prometheus and the SNS notifier, can be in private subnets.

**Comment by kostaleonard** [comment]
In the [eks-managed-node-group submodule](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/18.29.0/submodules/eks-managed-node-group), you can specify subnet IDs and node labels, e.g., "public" or "private". Then you could force pods to deploy to nodes with specific labels using a node selector.
